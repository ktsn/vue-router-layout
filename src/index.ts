import {
  VNode,
  Component,
  ConcreteComponent,
  h,
  defineComponent,
  shallowReactive,
} from 'vue'
import { RouteRecord, RouteLocationNormalized } from 'vue-router'

type Layout = { name: string; props: Record<string, unknown> }
type Optional<Object, Key extends keyof Object> = Omit<Object, Key> &
  Partial<Pick<Object, Key>>

function normalizeLayout(layout: any): Layout {
  if (typeof layout === 'string') {
    return {
      name: layout,
      props: {},
    }
  }

  if (layout && typeof layout === 'object' && 'name' in layout) {
    return {
      name: layout.name,
      props: layout?.props || {},
    }
  }

  throw new Error(
    `[vue-router-layout] Incorrect value for the 'layout' property.`
  )
}

/**
 * Find which layout the component should render.
 * If the component is not specified layout name, `default` is used.
 * Otherwise return undefined.
 */
function resolveLayout(matched: RouteRecord[]): Layout | undefined {
  const last = matched[matched.length - 1]
  if (!last) {
    return
  }

  const Component: any = last.components.default
  if (!Component) {
    return
  }

  const isAsync = typeof Component === 'function' && !Component.options
  if (isAsync) {
    return
  }

  const defaultLayout: Layout = { name: 'default', props: {} }

  return getLayout(Component) || defaultLayout
}

function getLayout(
  Component: any /* ComponentOptions | VueConstructor */
): Layout | undefined {
  const isCtor = typeof Component === 'function' && Component.options
  const options = isCtor ? Component.options : Component

  if (options.layout) {
    return normalizeLayout(options.layout)
  }

  // Retrieve super component and mixins
  const mixins: any[] = (options.mixins || []).slice().reverse()
  const extend: any = options.extends || []

  for (const c of mixins.concat(extend)) {
    const layout = getLayout(c)
    if (layout) {
      return layout
    }
  }
}

function loadAsyncComponents(route: RouteLocationNormalized): Promise<unknown> {
  const promises: Promise<unknown>[] = []

  route.matched.forEach((record) => {
    Object.keys(record.components).forEach((key) => {
      const component: any = record.components[key]
      const isAsync = typeof component === 'function' && !component.options

      if (isAsync) {
        promises.push(
          component().then((loaded: any) => {
            record.components[key] = normalizeEsModuleComponent(loaded)
          })
        )
      }
    })
  })

  return Promise.all(promises)
}

function normalizeEsModuleComponent(
  comp: Component | { default: Component }
): Component {
  const c: any = comp
  const isEsModule =
    c.__esModule ||
    (typeof Symbol !== 'undefined' && c[Symbol.toStringTag] === 'Module')
  return isEsModule ? c.default : c
}

function install() {
  console.info(
    '[vue-router-layout] app.use(VueRouterLayout) is no longer needed. You can safely remove it.'
  )
}

export function createRouterLayout(
  resolve: (layoutName: string) => Promise<Component | { default: Component }>
) {
  return defineComponent({
    name: 'RouterLayout',

    data() {
      return {
        layout: undefined as Layout | undefined,
        layouts: shallowReactive(
          Object.create(null) as Record<string, Component>
        ),
      }
    },

    async beforeRouteEnter(to, _from, next) {
      await loadAsyncComponents(to)

      const layout = resolveLayout(to.matched)
      const layoutComp = layout
        ? normalizeEsModuleComponent(await resolve(layout.name))
        : undefined

      next((vm: any) => {
        vm.layout = layout
        if (layout && layoutComp) {
          vm.layouts[layout.name] = layoutComp
        }
      })
    },

    async beforeRouteUpdate(to, _from, next) {
      try {
        await loadAsyncComponents(to)

        const layout = resolveLayout(to.matched) || this.layout
        if (layout && !this.layouts[layout.name]) {
          this.layouts[layout.name] = normalizeEsModuleComponent(
            await resolve(layout.name)
          )
        }

        this.layout = layout
        next()
      } catch (error) {
        next(error)
      }
    },

    render(): VNode | null {
      const layoutComponent = this.layout && this.layouts[this.layout.name]
      if (!layoutComponent) {
        return null
      }

      return h(layoutComponent as ConcreteComponent, {
        key: this.layout!.name,
        ...this.layout!.props,
      })
    },
  })
}

export default {
  install,
}

declare module '@vue/runtime-core' {
  interface ComponentCustomOptions {
    layout?: string | Optional<Layout, 'props'>
  }
}
