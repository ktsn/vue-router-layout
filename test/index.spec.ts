import { mount as _mount } from '@vue/test-utils'
import Vue, { Component, ComponentOptions } from 'vue'
import Router, { RouteConfig } from 'vue-router'
import { createRouterLayout } from '../src/index'

Vue.config.productionTip = false
Vue.config.devtools = false

Vue.use(Router)

const layouts: Record<string, Component> = {
  default: {
    template: '<div>Default <router-view/></div>',
  },

  foo: {
    template: '<div>Foo <router-view/></div>',
  },

  bar: {
    template: '<div>Bar <router-view/></div>',
  },
}

const RouterLayout = createRouterLayout((layout) => {
  return Promise.resolve(layouts[layout])
})

async function mount(children: RouteConfig[]) {
  const router = new Router({
    mode: 'abstract',
    routes: [
      {
        path: '/',
        component: RouterLayout,
        children,
      },
    ],
  })

  const Root = Vue.extend({
    template: '<router-view/>',
  })

  const wrapper = _mount(Root, {
    router,
  })
  router.push('/')
  await wrapper.vm.$nextTick()
  return wrapper
}

describe('RouterLayout component', () => {
  it('shows default layout', async () => {
    const Comp = {
      template: '<p>component</p>',
    }

    const wrapper = await mount([
      {
        path: '',
        component: Comp,
      },
    ])
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('shows named layout', async () => {
    const Comp = {
      layout: 'foo',
      template: '<p>component</p>',
    }

    const wrapper = await mount([
      {
        path: '',
        component: Comp,
      },
    ])
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("uses the leaf component's layout option", async () => {
    const Test1 = {
      layout: 'foo',
      template: '<div>Test1 <router-view/></div>',
    }

    const Test2 = {
      layout: 'bar',
      template: '<p>Test2</p>',
    }

    const wrapper = await mount([
      {
        path: '',
        component: Test1,
        children: [
          {
            path: '',
            component: Test2,
          },
        ],
      },
    ])
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('updates layout when route is changed', async () => {
    const Test1 = {
      layout: 'foo',
      template: '<p>Test1</p>',
    }

    const Test2 = {
      layout: 'bar',
      template: '<p>Test2</p>',
    }

    const wrapper = await mount([
      {
        path: '',
        component: Test1,
      },
      {
        path: 'test',
        component: Test2,
      },
    ])
    wrapper.vm.$router.push('/test')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('works with component constructor', async () => {
    const Test1 = Vue.extend({
      layout: 'foo',
      template: '<p>Test1</p>',
    })

    const Test2 = Vue.extend({
      template: '<p>Test2</p>',
    })

    const wrapper = await mount([
      {
        path: '',
        component: Test1,
      },
      {
        path: 'test',
        component: Test2,
      },
    ])
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.vm.$router.push('/test')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('works with async component', async (done) => {
    const Test1 = () => {
      return Promise.resolve({
        layout: 'foo',
        template: '<p>Test1</p>',
      })
    }

    const Test2 = () => {
      return new Promise<ComponentOptions<Vue>>((resolve) => {
        setTimeout(() => {
          resolve({
            layout: 'bar',
            template: '<p>Test2</p>',
          })
        }, 100)
      })
    }

    const wrapper = await mount([
      {
        path: '',
        component: Test1,
      },
      {
        path: 'test',
        component: Test2,
      },
    ])

    await wrapper.vm.$nextTick()

    // Foo - Test1
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.vm.$router.push('/test')

    // Foo - Test1 (Should not change layout until the async component is resolved)
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()

    setTimeout(() => {
      // Bar - Test2 (Should update layout after async component is resolved)
      expect(wrapper.html()).toMatchSnapshot()
      done()
    }, 200)
  })

  it('works with async component with default export', async (done) => {
    const Test1 = () => {
      return Promise.resolve({
        layout: 'foo',
        template: '<p>Test1</p>',
      })
    }

    const Test2 = () => {
      return new Promise<any>((resolve) => {
        setTimeout(() => {
          import('./dummy').then(resolve)
        }, 100)
      })
    }

    const wrapper = await mount([
      {
        path: '',
        component: Test1,
      },
      {
        path: 'test',
        component: Test2,
      },
    ])

    await wrapper.vm.$nextTick()

    // Foo - Test1
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.vm.$router.push('/test')

    // Foo - Test1 (Should not change layout until the async component is resolved)
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()

    setTimeout(() => {
      // Bar - Test2 (Should update layout after async component is resolved)
      expect(wrapper.html()).toMatchSnapshot()
      done()
    }, 200)
  })

  it('should not instantiate destination page until layout is rendered', async (done) => {
    const created = jest.fn()

    const Test1 = () => {
      return Promise.resolve({
        layout: 'foo',
        template: '<p>Test1</p>',
      })
    }

    const Test2 = () => {
      return new Promise<any>((resolve) => {
        setTimeout(() => {
          resolve({
            layout: 'bar',
            template: '<p>Test2</p>',
            created,
          })
        }, 100)
      })
    }

    const wrapper = await mount([
      {
        path: '',
        component: Test1,
      },
      {
        path: 'test',
        component: Test2,
      },
    ])

    wrapper.vm.$router.push('/test')

    setTimeout(() => {
      expect(created).toHaveBeenCalledTimes(1)
      done()
    }, 200)
  })

  it('pulls layout value from extends', async () => {
    const Super = {
      layout: 'foo',
    }

    const Comp = {
      extends: Super,
      template: '<p>component</p>',
    }

    const wrapper = await mount([
      {
        path: '',
        component: Comp,
      },
    ])
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('inherits layout value from super constructor', async () => {
    const Super = Vue.extend({
      layout: 'foo',
    })

    const Comp = Super.extend({
      template: '<p>component</p>',
    })

    const wrapper = await mount([
      {
        path: '',
        component: Comp,
      },
    ])
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('prioritizes mixins option than extends', async () => {
    const Mixin = {
      layout: 'bar',
    }

    const Super = {
      layout: 'foo',
    }

    const Comp = {
      mixins: [Mixin],
      extends: Super,
      template: '<p>component</p>',
    }

    const wrapper = await mount([
      {
        path: '',
        component: Comp,
      },
    ])
    expect(wrapper.html()).toMatchSnapshot()
  })
})
