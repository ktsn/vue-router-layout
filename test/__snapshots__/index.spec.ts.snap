// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterLayout component inherits layout value from super constructor 1`] = `"<div>Foo <p>component</p></div>"`;

exports[`RouterLayout component prioritizes mixins option than extends 1`] = `"<div>Bar <p>component</p></div>"`;

exports[`RouterLayout component pulls layout value from extends 1`] = `"<div>Foo <p>component</p></div>"`;

exports[`RouterLayout component shows default layout 1`] = `"<div>Default <p>component</p></div>"`;

exports[`RouterLayout component shows named layout 1`] = `"<div>Foo <p>component</p></div>"`;

exports[`RouterLayout component updates layout when route is changed 1`] = `"<div>Bar <p>Test2</p></div>"`;

exports[`RouterLayout component uses the leaf component's layout option 1`] = `"<div>Bar <div>Test1 <p>Test2</p></div></div>"`;

exports[`RouterLayout component works with async component 1`] = `"<div>Foo <p>Test1</p></div>"`;

exports[`RouterLayout component works with async component 2`] = `"<div>Foo <p>Test1</p></div>"`;

exports[`RouterLayout component works with async component 3`] = `"<div>Bar <p>Test2</p></div>"`;

exports[`RouterLayout component works with component constructor 1`] = `"<div>Foo <p>Test1</p></div>"`;

exports[`RouterLayout component works with component constructor 2`] = `"<div>Default <p>Test2</p></div>"`;
